Format Java source files automatically

pugnascotia opened the pull request:
This PR introduces support for formatting Java code using ~~[google-java-format](https://github.com/google/google-java-format)~~ [Eclipse JDT](https://www.eclipse.org/jdt/). Fomatting can be verified or executed out through Gradle. There is a CLI, and plugins for Eclipse and Jetbrains IDEs.

Formatting is executed with:

    ./gradlew spotlessApply

Checking is executed with:

    ./gradlew spotlessJavaCheck

Checking is also carried out during the `precommit` task.

I've updated `CONTRIBUTING.md` with some notes on formatting, IDE plugins, etc.

Sub-project needs to opt-in before the formatter will look at them - currently the list of opted-in projects is empty. This PR just adds the mechanism.

pugnascotia committed:
Format Java using google-java-format

Fomatting can be checked or carried out through Gradle.

pugnascotia committed:
Add notes on handling file that do not format

Also fix the interleaving of imports and comments in a few files, which
the formatted rejected.

elasticmachine commented:
Pinging @elastic/es-core-infra

atorok commented on CONTRIBUTING.md:
```diff
@@ -155,19 +155,47 @@ For Eclipse, go to `Preferences->Java->Installed JREs` and add `-ea` to
+     '*'` and `Names count to use static import with '*'`. Set their values
+     to 99999 or some other absurdly high value.
+
+If a file compiles successfully but fails to format, the Gradle plugin does
```
I fully support this initiative, but see this as a huge  disappointment. 
We absolutely have to address this if we were to merge either by using some other tool, doing what is described here automatically or just fixing the plugin.

atorok commented on build.gradle:
```diff
@@ -99,6 +104,11 @@ subprojects {
+  // Although the plugin is applied at the top-level, that doesn't make
+  // it available on subprojects.
+  project.apply plugin: 'com.github.sherter.google-java-format'
+  check.dependsOn 'verifyGoogleJavaFormat'
```
`precommit.check` should have worked  in the dame way, I do think we need that there.

pugnascotia replied:
```suggestion
  precommit.check.dependsOn 'verifyGoogleJavaFormat'
```

I tried the above but Gradle didn't like it. I'll keep looking it at, all suggestions welcome.

alpar-t reviewed (commented):
This is great ! Thanks for working on it ! 

I added some comments, glad to help with any of them. 

Does the tool re-format `false == something`  to  ` ! something ` ? 
We feel strongly for the former and most tool would re-write that. 
IDEA constantly complains about that with no apparent way to turn it off too.

atorok commented:
PS: a few other examples of the what the new format changes would be useful too. Maybe add some files that have most changes ? 

PS2: How long does it take for it to run ? Looking to see how much longer `precommit` would take.  Can it maybe replace what `checkstyle` is doing ?

pugnascotia committed:
Apply format plugin in the right place

pugnascotia committed:
Some big examples of reformatted code

pugnascotia commented:
The formatter is quick, I can reformat the entire codebase in about 4-5 minutes on my laptop. The IDE plugin formats a file instantaneously. 

I've added somne choice reformatted examples. The formatting is actually quite conservative, and seems mostly concerned with whitespace, line length and ordering imports. From that perspective, it is nowhere near to replacing Checkstyle.

atorok commented:
How long does it take to check that the format is the one we expect ?

atorok commented on CONTRIBUTING.md:
```diff
@@ -155,19 +155,47 @@ For Eclipse, go to `Preferences->Java->Installed JREs` and add `-ea` to
+[google-java-format](https://github.com/google/google-java-format), which
+enforces the [Google Java
+Style](https://google.github.io/styleguide/javaguide.html). There are
+plugins available for Eclipse, JetBrains IDEs such as IntelliJ IDEA, and a
```
Could we link directly to the plugins ? 
Alternatively @mark-vieira do you know a way to generate an IDEA configuration that would at least suggests users to install the plugin ?

mark-vieira replied:
I'm not aware of any. I suspect something like installing a plugin is not something most folks want to happen when importing a project. There are likely security concerns there.

pugnascotia committed:
Switch to Spotless for code formatting

The preview gradle integration for google-java-format did an inadequate
job at reporting problems, so I've switched to using Spotless. I had to
upgrade the JGit dependency in order to get error reporting to work.

Also commit formatting changes to `TestingConventionsTasks.java`, since
it demonstrates a file that had suffered from being formatted.

pugnascotia commented:
## Spotless

I had another go at using `spotless`, and this time I got it to work. I don’t know what my problem was before, and I checked that Spotless does report errors arising from `google-java-format` (you get a great big stack trace, but the error is there). For plain old formatting violations, it prints a summary / report of the problems. I managed to bind it to `precommit` as well.

Note that I had to upgrade a `jgit` dependency, otherwise Spotless would call a non-existent `jgit` method when trying to print a diff and explode.

### Groovy / Gradle

I also tried formatting the Groovy and Gradle files with spotless, which almost worked great, but it relies on an Eclipse formatter, and it did some very strange things with indentation, so I removed it again. One for the future perhaps.

## Formatting notes

I also found a pathological formatting case in `buildSrc/src/main/java/org/elasticsearch/gradle/precommit/TestingConventionsTasks.java`. This file has some deeply nested expressions, which end up pretty unreadable under `google-java-format` with 4-space indents.The formatter adds a double indent when a statement wraps onto another line, which is mostly OK-ish, but in this file’s case, the code ends up smeared across the screen, so much so that it triggers the Checkstyle line-length warning. I committed the file to demonstrate.

Now, to address the immediate issue of Checkstyle, we could disable that check for Java files, because the formatter (usually) doesn’t allow long lines. That aside, if we proceed with formatting the codebase then we may want to follow up with some refactoring to make such files more palatable (e.g. extract expressions to vars). The other options is to switch to 2-space indents, which would probably reduce the size of diff significantly (if you ignore whitespace changes). What do people think?

On a similar note, the formatter will wrap comments to satisfy the line length constraint, which sometimes mangles the readability of comments somewhat. It’s not a big deal, and perhaps the team can fix up any particularly bad examples as they find them.

## Issues with JavaDoc

Entertainingly, the formatter tries much harder to do something pretty with JavaDoc comments, and will even format HTML (and remove some closing HTML tags in the process, which I don’t agree with). Less entertainingly, the formatter will sometimes break a line after a `<code>` tag containing e.g. `@something`, which Java then misinterprets as a non-existent annotation and breaks the build. This can be worked around by editing the JavaDoc to cause the line to break elsewhere. There is an issue for this, and I left a comment:

https://github.com/google/google-java-format/issues/84

atorok commented on build.gradle:
```diff
@@ -98,6 +99,23 @@ subprojects {
+      }
+    }
+
+    project.afterEvaluate {
```
The after evaluate should not be needed here as `precommit` should exist at this time and the argument to `dependsOn` is a string that gets resolved lather.

atorok commented on build.gradle:
```diff
@@ -98,6 +99,23 @@ subprojects {
+        googleJavaFormat().aosp()
+        // Required for :modules:lang-painless: - avoids a rule failing to
+        // make up its mind about formatting.
+        paddedCell()
```
could perhaps this go into that specific project?

atorok commented:
This  looks good to me. 
We'll for sure get some oddly formatted files that we'll have to work on, 
or maybe reformat as we discover them, but that's a fact of life that we can run into 
Today, it's not a problem necessarily created by a format er ( even trough it could be aggravated ) 

The only problem I see is that making this happen all at once will be very difficult without generating lots of conflicts in all directions. 
It would be nice if we had a way to do this on a project - by project or in some cases package by package basis to make it less disruptive. 

E.x. we could specifically enable this for projects until we have them all and then enable for all by default ? Or enable for all and specifically disable for all existing project so new projects added will have it enabled by default and existing ones will migrate.

pugnascotia committed:
Changes to make Spotless format consistently

pugnascotia committed:
Only format projects on an accept list.

pugnascotia committed:
Revert formatted outside of buildSrc

pugnascotia committed:
Format buildSrc directory

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into auto-format-java-source

pugnascotia committed:
Reformat after merge

pugnascotia commented:
@atorok I would have sworn I needed the `afterEvaluate`, but I took it out and it works 🤷‍♂

I've added a list of projects to format, and included only `:build-tools`. We could add projects to the list gradually and avoid dropping a giant conflict bomb on the team.

I also investigating the `paddedCell()` stuff, and wrote up some notes in the contributing guide. The issues were largely to do with comments and JavaDoc, and I fixed the issues that I found.

@rjernst @jasontedor Feels like this PR is in good shape, and the outstanding questions are:

   - Does Core/Infra want to recommend this path to the wider team? (I can do this if we're happy)
   - Do we want to keep 4-space indents? 😛

pugnascotia committed:
Update checkstyle suppressions

atorok commented on buildSrc/src/main/java/org/elasticsearch/gradle/BwcVersions.java:
```diff
@@ -34,59 +37,59 @@
-    private static final Pattern LINE_PATTERN = Pattern.compile(
-        "\\W+public static final Version V_(\\d+)_(\\d+)_(\\d+)(_alpha\\d+|_beta\\d+|_rc\\d+)? .*"
-    );
+    private static final Pattern LINE_PATTERN =
```
I think this is an example where the formatter had no business and made things worse.

pugnascotia replied:
I don't exactly dislike what GJF did, but it's interesting that e.g. Prettier formats code more like we do at the moment. I wonder how far [prettier-java](https://github.com/jhipster/prettier-java) will get - not that I want to add a JS build dependency to a Java project 😅

atorok commented on buildSrc/src/test/java/org/elasticsearch/gradle/precommit/DependencyLicensesTaskTests.java:
```diff
@@ -57,16 +56,19 @@ public void givenProjectWithLicensesDirButNoDependenciesThenShouldThrowException
 
     @Test
-    public void givenProjectWithoutLicensesDirNorDependenciesThenShouldReturnSilently() throws Exception {
+    public void givenProjectWithoutLicensesDirNorDependenciesThenShouldReturnSilently()
```
Do we have a different line length setting in the formatter than in checkstyle ? 
Not sure why this line was changed

pugnascotia replied:
GJF has a line length of 100, as opposed to 140 for Checkstyle.

jasontedor replied:
That's an item that we were considering forking `google-java-format` for, not because we necessarily think that 140 is the right line length (I don't), but because we wanted to avoid a *massive* reformatting of the codebase.

mark-vieira commented:
> Now, to address the immediate issue of Checkstyle

I'm not sure. We have some pretty specific Checkstyle rules that I'm not sure there are equivalents in spotless. [For example](https://github.com/elastic/elasticsearch/blob/ae5919b2c8c30ea0be5b8f198ea7cda7aabce0bf/buildSrc/src/main/resources/checkstyle.xml#L21-L32).

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/ConcatFilesTask.java:
```diff
@@ -33,9 +32,7 @@
-/**
- * Concatenates a list of files into one and removes duplicate lines.
- */
+/** Concatenates a list of files into one and removes duplicate lines. */
```
I must say, I'm not a fan of these single line multi-line comments.

mark-vieira commented:
> We'll for sure get some oddly formatted files that we'll have to work on,
or maybe reformat as we discover them

We don't want to be in a position where we are constantly battling the formatter. If it implements a code style we simply refuse to accept that's a problem as we'll keep introducing it.

mark-vieira commented on build.gradle:
```diff
@@ -98,6 +99,30 @@ subprojects {
+          googleJavaFormat().aosp()
+
+          // See CONTRIBUTING.md for details of when to enabled this.
+          // paddedCell()
```
Can we perhaps put this behind a flag? Say `-Dspotless.paddedCell` or something. I like to avoid toggles that involve commenting or uncommenting stuff because it's very easy for that stuff to be mistakenly committed to source control.

jasontedor commented:
>>We'll for sure get some oddly formatted files that we'll have to work on, or maybe reformat as we discover them

>We don't want to be in a position where we are constantly battling the formatter. If it implements a code style we simply refuse to accept that's a problem as we'll keep introducing it.

We definitely don't want to fight the formatter. The idea has been that we would accept whatever the formatter tells us is *the* format, it's intended to end all formatting arguments once and for all (unless we vehemently disagree with something that it does, in which case we should fork or not use the formatter but we also want to minimize the number of forks that we make). To my previous point, the only one that we didn't want to relax is 140 columns, only because of the implications for re-formatting an incredibly large fraction of the codebase.

pugnascotia committed:
Add an option to enable paddedCell() from the gradle CLI

pugnascotia commented:
I'm going to take a look at using the Eclipse formatter as an alternative, since it's easy to call from Spotless. It does mean that I need to configure the formatting settings, but thereafter we can just apply them, and it's possible to import them into both Eclipse and IntelliJ. Apparently you can [call spotless on specific files](https://github.com/diffplug/spotless/tree/master/plugin-gradle#can-i-apply-spotless-to-specific-files) too.

pugnascotia committed:
Format java with with Eclipse JDT instead

pugnascotia committed:
Reformat buildSrc/

pugnascotia commented:
Right, I've switched the branch to format with Eclipse. Please take a look.

Pros:
   * We have a lot more control over the format, such as line length

Cons:
   * We have a lot more control over the format
   * The formatting result is probably more invasive overall, but closer to what we'd chose (I think?). I can continue to tinker with the settings if we think it would help.

atorok commented:
We will have to consider how this affects backporting. 
If we run the tool both on master and 7.x will git be able to correctly track the changes and not make our life impossible when back-porting ? 
I'm not saying it will, just saying we need to give it more thought.

pugnascotia committed:
More formatter settings tweaks

mark-vieira commented:
> We will have to consider how this affects backporting.
If we run the tool both on master and 7.x will git be able to correctly track the changes and not make our life impossible when back-porting ?

This is a good point. The assumption would be that if we introduce this, we would first reformat all existing code. Then all subsequent PRs should be on top of that.

That said, we'd have to implement this back to all branches we intend to backport changes to, and also reformat those branches, or else as Alpar points out, backporting PRs from post-formatted branches is going to be a nightmare.

jasontedor commented:
>This is a good point. The assumption would be that if we introduce this, we would first reformat all existing code. Then all subsequent PRs should be on top of that.

>That said, we'd have to implement this back to all branches we intend to backport changes to, and also reformat those branches, or else as Alpar points out, backporting PRs from post-formatted branches is going to be a nightmare.

It's for this reason that when we discussed this in the past we concluded the best time to make a transition like this is when we cut a major branch from master. If we did this here, we would hold this change until right before we cut 8.0 and 8.x from master. In this way, we would end up with 8.0, and 8.x, and master all formatted the same, and then backports from master to 8.x and 8.0, and 8.x to 8.0 would go through without pain since the branches can not have deviated between formatting master and cutting the branches. The only wrinkle is 7.last but I find that as time goes on (e.g., from master to 6.8) such backports become less and less clean anyway.

pugnascotia commented:
So just so confirm, in that case we wouldn't gradually opt in Gradle projects over time, but instead format the whole repo just before cutting the branches. Correct? If so, I think I should send an email about this to the broader ES team sooner rather than later.

(I'm also looking for further feedback, if you have any, on the formatting itself, given the switch to the Eclipse JDT formatter.)

pugnascotia committed:
Revert formatted ahead of merge with master

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into auto-format-java-source

pugnascotia committed:
Leave the formatting project list empty, so we can merge the Gradle changes

pugnascotia committed:
Update CONTRIBUTING.md

pugnascotia commented:
I've trimmed this PR now so that it can be merged without doing anything 😁

pugnascotia committed:
Update CONTRIBUTING.md

rjernst commented on build.gradle:
```diff
@@ -98,6 +99,35 @@ subprojects {
+      spotless {
+        java {
+
+          removeUnusedImports()
```
Why are these unused imports or trailing whitespace settings not in the xml file?

pugnascotia replied:
The Eclipse formatter doesn't perform all these operations. That said, I can (and have) removed the `endWithNewline()` option.

rjernst commented on build.gradle:
```diff
@@ -98,6 +99,35 @@ subprojects {
+
+          // See CONTRIBUTING.md for details of when to enabled this.
+          if (System.getProperty('spotless.paddedcell') != null) {
+            paddedCell()
```
Shouldn't we have everything about the formatting doable by the IDE plugins?

pugnascotia replied:
I don't think we'll have much need to enable this on a day-to-day basis, and it's outside the remit of the IDE plugins. This option lets Spotless cope when formatting isn't idempotent, which happened in a few places with GJF. It's a seperate thing from the configured formatting options.

rjernst reviewed (commented):
Could you please change the title now that we are no longer evaluating google-java-format? I left a couple minor comments.

pugnascotia committed:
Remove spotless command that the formatter already handles

atorok commented:
@pugnascotia hope you don't mint that I back-ported this to `7.x` on of my other back-ports wasn't clean because of this one

pugnascotia commented:
I don't mind at all, we may very well end up formatting `7.x` anyway.

atorok commented:
Even if we don't, there's value in keeping the build code similar across branches to make it easy to pack-port improvements.
